source/common/memory/debug.cc

```diff
@@ -0,0 +1,118 @@
+// Very simple memory debugging overrides for operator new/delete, to
+// help us quickly find simple memory violations:
+//   1. Double destruct
+//   2. Read before write (via scribbling)
+//   3. Read after delete (via scribbling)
+//
+// Note that valgrind does all of this much better, but is too slow to run all
+// the time. asan does read-after-delete detection but not read-before-init
+// detection. See
+// https://clang.llvm.org/docs/AddressSanitizer.html#initialization-order-checking
+// for more details.
+
+// Principle of operation: add 8 bytes to every allocation. The first
+// 4 bytes are a marker (LiveMarker1 or DeadMarker1). The next 4
+// bytes are used to store size of the allocation, which helps us
+// know how many bytes to scribble when we free.
+//
+// This code was adapted from mod_pagespeed, and adapted for Envoy
+// style. Original source:
+// https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/base/mem_debug.cc
+
+// We keep a global count of bytes allocated so that memory-consumption tests
+// work with memory debugging. Put another way, if we disable tcmalloc when
+// compiling for debug, we want the memory-debugging tests to work, otherwise we
+// can't debug them.
+#include "common/memory/debug.h"
+
+#include <atomic>
+#include <cassert> // don't use Envoy ASSERT as it may allocate memory.
+#include <cstdlib>
+
+#include "common/memory/align.h"
+
+static std::atomic<int64_t> bytes_allocated(0);
```

Comment:
Could we also track `allocated_bytes_with_overhead` to account for alignment and allocation headers?